docs/source/app-dev/app-arch.rst

```diff
@@ -246,3 +235,79 @@ There are some other identifiers that are determined by your client code. These
    :width: 6.5in
    :height: 3.69444in
+
+Testing
+=======
+
+Testing is fundamental to ensure correctness and favor maintainability.
+
+Testing is usually divided in different categories according to its scope and aim:
+
+- unit testing verifies single properties of individual components
+- integration testing verifies that an aggregation of components behaves as expected
+- acceptance testing checks that the overall behavior of a whole system satisfies certain criteria
+
+Both tests in the small scale (unit testing) and large (acceptance testing) tend to be specific to the given component or system under test.
+
+This chapter focuses on providing portable approaches and techniques to perform integration testing between your components and an actual running ledger.
+
+Test the business logic with a ledger
+*************************************
+
+In production, your application is going to interact with a DAML model deployed on an actual ledger.
+
+Each model is usually specific to a business need and describes specific protocols.
+
+Mocking a ledger response is usually not desirable to test the business logic, because so much of it is encapsulated in the DAML model.
+
+This makes integration testing with an actual running ledger fundamental to evaluating the correctness of an application.
+
+This is usually achieved by running a ledger as part of the test process and run several tests against it, possibly coordinated by a test framework.
+
+Since the in-memory sandbox shipped as part of the SDK is a full-fledged implementation of a DAML ledger, it's usually the tool of choice for these tests.
+
+Please note that this does not replace acceptance tests with the actual ledger implementation that your application aims to use in production.
+
+Whatever your choice is, sharing a single ledger to run several tests is a suggested best practice.
+
+Share the ledger
+****************
+
+Booting a ledger and loading DAML code into it takes time. The shorter your tests, the more this overhead is going to impact the running time of the tests.
+
+Given the fundamental need of working against a running ledger to verify the business logic, it's quite common to have integration test that, in appearance, resemble unit tests, in that they check an individual property of a component you wrote, although in an integration test setting.
+
+This means it's quite common to write short tests, which aim, just like unit tests, to give quick feedback to the developer.
+
+In order for this to happen, saving on the need to boot a ledger multiple times can dramatically decrease the time necessary to run a suite of tests.
+
+Since your tests are going to share the same ledger, they must be designed to not interfere with each other.
+
+Design your tests to share the ledger
+*************************************
+
+Both the transaction and the active contract service offer the possibility of filtering by party. Parties can thus be used as a way to isolate tests.
```

Comment:
I think using disjoint sets of parties for each test is exactly the right approach, that's how Truffle does it.
However, your description here relies on starting with no active contracts. Getting the ledger into a desired state using automation is quite cumbersome. I often use scenarios. I think it'd be good to give some advice on ledger initialization.

Reply:
Scenarios can run only on a clean ledger, which would require to either disqualify the whole idea of sharing a ledger or need to run with a scenario that loads all possible events. Furthermore, this approach is incompatible with the suggested approach of allocating new parties for each test. Rather, one thing I left out and that greatly improves setting up the desired state is leveraging codegen where possible, which I could add as a suggestion.

Reply:
Yeah, with scenarios you'd have to define your disjoint sets of parties in each scenario. Scenarios are composable so you can just write a container-scenario that runs the setup for each test .

Reply:
The problem in using this approach is having to manually keep synchronized an ever-growing list of users that have to carefully be created on both sides for this to work. I would favor using the `PartyManagementService` along with good helpers and possibly codegen to make the setup as lightweight as possible. This approach has been taken when developing the new integration tests and setting up a test scenario is very much not cumbersome ([example](https://github.com/digital-asset/daml/blob/master/ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Divulgence.scala#L18-L22)). I can include your suggestion but it's important to highlight the possible maintenance burden for large test suites.

Reply:
After a bit of thought and interesting input from @gerolf-da, it's probably the case to reinforce the idea of having expressive bindings and possibly code generation do the heavy lifting for you of coming to a state.